pyvrp/cpp/repair/greedy_repair.cpp

```diff
@@ -0,0 +1,164 @@
+void setupRoutes(Clients &clients,
+                 Routes &routes,
+                 SolRoutes const &solRoutes,
+                 ProblemData const &data)
+{
+    assert(clients.empty() && routes.empty());
+
+    // Doing this avoids re-allocations, which would break the pointer structure
+    // that Route and Route::Node use.
+    clients.reserve(data.numClients() + 1);
+    routes.reserve(solRoutes.size());
+
+    for (size_t client = 0; client <= data.numClients(); ++client)
+        clients.emplace_back(client);
+
+    size_t idx = 0;
+    for (auto const &solRoute : solRoutes)
+    {
+        auto &route = routes.emplace_back(data, idx++, solRoute.vehicleType());
+        for (auto const client : solRoute)
+            route.push_back(&clients[client]);
+
+        route.update();
+    }
+}
+
+// Turns the given routes into a solution.
+Solution exportRoutes(ProblemData const &data, Routes const &routes)
+{
+    std::vector<Solution::Route> solRoutes;
+    solRoutes.reserve(routes.size());
+
+    for (auto const &route : routes)
+    {
+        std::vector<size_t> visits;
+        visits.reserve(route.size());
+
+        for (auto *node : route)
+            visits.push_back(node->client());
+
+        solRoutes.emplace_back(data, visits, route.vehicleType());
+    }
+
+    return {data, solRoutes};
+}
+
+// Evaluates the move of inserting U after V.
+pyvrp::Cost evaluate(Route::Node *U,
+                     Route::Node *V,
+                     ProblemData const &data,
+                     CostEvaluator const &costEvaluator)
+{
+    assert(!U->route() && V->route());
+    auto *route = V->route();
+
+    auto const deltaDist = data.dist(V->client(), U->client())
+                           + data.dist(U->client(), n(V)->client())
+                           - data.dist(V->client(), n(V)->client());
+
+    auto const &uClient = data.client(U->client());
+    auto deltaCost = static_cast<pyvrp::Cost>(deltaDist) - uClient.prize;
+
+    deltaCost += costEvaluator.loadPenalty(route->load() + uClient.demand,
+                                           route->capacity());
+    deltaCost -= costEvaluator.loadPenalty(route->load(), route->capacity());
+
+    // If this is true, adding U cannot decrease time warp in V's route enough
+    // to offset the deltaCost.
+    if (deltaCost >= costEvaluator.twPenalty(route->timeWarp()))
+        return deltaCost;
+
+    auto const vTWS = TWS::merge(data.durationMatrix(),
+                                 route->twsBefore(V->idx()),
+                                 TWS(U->client(), uClient),
+                                 route->twsAfter(V->idx() + 1));
+
+    deltaCost += costEvaluator.twPenalty(vTWS.totalTimeWarp());
+    deltaCost -= costEvaluator.twPenalty(route->timeWarp());
+
+    return deltaCost;
+}
+}  // namespace
+
+Solution pyvrp::repair::greedyRepair(Solution const &solution,
+                                     DynamicBitset const &unplanned,
```

Comment:
I think we should use `std::vector` here instead of a DynamicBitset.
- Vector preserves the ordering of clients: this may impact the effectiveness of repair procedures. The current implementation always inserts clients from 1 to n.
- We're not really using the set operations anyway in greedy repair, and I also don't think there will be a use case for it in the future.